### PR TITLE
keep method when cloning Request, fixes #36522

### LINKED ISSUE #36522: creating a new Request from an existing object does not keep the request's method
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.13.2
      npm: 8.7.0
      Yarn: 1.22.15
      pnpm: 6.11.0
    Relevant packages:
      next: 12.1.5
      react: 17.0.2
      react-dom: 17.0.2

### What browser are you using? (if relevant)

n/a

### How are you deploying your application? (if relevant)

n/a

### Describe the Bug

1. when creating a Request object from an existing Request object using the polyfill, the method is not copied over

```js
const request = new Request('http://example.com', { method: 'POST', body: 'hello'})
const request2  = new Request(request)
```

throws: `TypeError: Request with GET/HEAD method cannot have body`

this is because the polyfill takes method from the optional `init` parameter of Request#constructor: https://github.com/vercel/next.js/blob/842e4ec02381234bdea82a9c95e75cc79e211a09/packages/next/server/web/spec-compliant/request.ts#L19-L26

2. when fetching a Request object in a _middleware, the method is ignored:

```js
const request = new Request('http://example.com', { method: 'POST', body: 'hello'})
const response = await fetch(request)

expect(response.req.method).toEqual('POST') // <-- return 'GET'
```

https://github.com/vercel/next.js/blob/d5e767b20d06dfd625f70267b713af52917e5897/packages/next/server/web/sandbox/context.ts#L152-L173

### Expected Behavior

the expected behaviour is the `method` and any other properties are copied over to the new request object

### To Reproduce

for example in _middleware.ts:

```js
const request = new Request('http://example.com', { method: 'POST', body: 'hello'})
const request2  = new Request(request)
request2.method === 'POST'
```

### PR DESCRIPTION
**Pull request by gazs**
This fixes #36522 by copying the method from the input Request object when constructing a new one. A new unit test is added to verify this.

Fixes: #36522 

## Bug

- [x] Related issues linked using `fixes #36522`
- [n/a] Integration tests added
- [n/a] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [x] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| buildDuration | 15.3s | 15.3s | ⚠️ +11ms |
| buildDurationCached | 6.1s | 6.1s | ⚠️ +32ms |
| nodeModulesSize | 463 MB | 463 MB | ⚠️ +245 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.02 | 3 | -0.02 |
| / avg req/sec | 827.69 | 833.3 | +5.61 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.172 | 1.174 | 0 |
| /error-in-render avg req/sec | 2133.15 | 2129.88 | ⚠️ -3.27 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.71 kB | 5.71 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.64 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.3 kB | 16.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| buildDuration | 17.2s | 17.3s | ⚠️ +44ms |
| buildDurationCached | 6.1s | 6.1s | ⚠️ +39ms |
| nodeModulesSize | 463 MB | 463 MB | ⚠️ +245 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.036 | 3.028 | -0.01 |
| / avg req/sec | 823.34 | 825.63 | +2.29 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.199 | 1.158 | -0.04 |
| /error-in-render avg req/sec | 2084.62 | 2158.92 | +74.3 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.1 kB | 29.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.76 kB | 2.76 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.5 kB | 16.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | gazs/next.js request-clone-keep-method | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: ccaeb9ad4941c676b487994852c57afab467a356

File: packages/next/server/web/spec-compliant/request.ts [hunk 1]
```diff
@@ -16,7 +16,9 @@ class BaseRequest extends Body implements Request {
   }
 
   constructor(input: BaseRequest | string, init: RequestInit = {}) {
-    const method = init.method?.toUpperCase() ?? 'GET'
+    const method =
+      init.method?.toUpperCase() ??
+      (input instanceof BaseRequest ? input.method?.toUpperCase() : 'GET')
```
Comment (SukkaW) [inline packages/next/server/web/spec-compliant/request.ts:21]:
```suggestion
    const method = init.method?.toUpperCase() ?? 
    (typeof input !== 'string' ? input.method?.toUpperCase() : 'GET')
```

Comment (ijjk) [reply packages/next/server/web/spec-compliant/request.ts:21]: Would we also need to update the checks to consider `input` being nullish if we check for `!== 'string'` here??

**Review by ijjk — APPROVED** [review]
Thanks for the PR!

**Pull request merged**